synapse/rest/admin/federation.py

```diff
@@ -0,0 +1,96 @@
+# Copyright 2021 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+from http import HTTPStatus
+from typing import TYPE_CHECKING, Tuple
+
+from synapse.api.errors import Codes, SynapseError
+from synapse.http.servlet import RestServlet, parse_integer, parse_string
+from synapse.http.site import SynapseRequest
+from synapse.rest.admin._base import admin_patterns, assert_requester_is_admin
+from synapse.storage.databases.main.transactions import DestinationSortOrder
+from synapse.types import JsonDict
+
+if TYPE_CHECKING:
+    from synapse.server import HomeServer
+
+logger = logging.getLogger(__name__)
+
+
+class FederationDestinationRestServlet(RestServlet):
+    """Get request to list all destinations.
+    This needs user to have administrator access in Synapse.
+
+    GET /_synapse/admin/v1/federation/destinations?from=0&limit=10
+
+    returns:
+        200 OK with list of destinations if success otherwise an error.
+
+    The parameters `from` and `limit` are required only for pagination.
+    By default, a `limit` of 100 is used.
+    The parameter `destination` can be used to filter by destination.
```

Comment:
I wonder if this might be served better by a separate endpoint for querying the status of one specific destination?

Reply:
In the future, an API for a single destination can certainly be added. However, the aim is to design the API in a similar way like rooms or users.

Reply:
I agree with @DMRobertson that this sounds like it should be a separate API, maybe `/federation/destinations/{destination}$`? That feels like the way rooms are handled currently.

Reply:
I had in mind the use cases https://github.com/matrix-org/synapse/issues/7982#issue-668094581 outlined here. 

But I don't mind there being a generic means to answer "what's the state of federation on my homeserver?", and this seems like a start to that. It's flagged as `experimental and "subject to change"` in the docs so I think we should get this in, see if it's helpful to people and iterate from there.

Reply:
It is not easy for me to work on it. Information on the subject is hard to find.